### ISSUE TITLE
Apply jacobian on wrong variable.

### ISSUE THREAD

**Issue by CZ-Wu** [SuLvXiangXin/zipnerf-pytorch#15]
Hi, thanks for your great work. I am reading your code and have found some problems with your implementation. From Eq (8) of Zip-NeRF and Eq (8) of Mip-NeRF 360, jacobian is applied on sampled Gaussian mean before contraction. However, in your [first commit](https://github.com/SuLvXiangXin/zipnerf-pytorch/commit/e821b523d63d98f293a75852350b302294231e6f#diff-c910b384e82ad342cd30d9a2739890a58d8274d83c8010d5994f090bf8193534) (the code below), you apply jacobian on the mean after contraction, though you have updated your code in latest commit which is much faster, I have tested that it makes no difference.

```
    pre_shape = mean.shape[:-1]
    mean = mean.reshape(-1, 3)
    std = std.reshape(-1)
    mean = contract(mean)
    jvp = vmap(jacrev(fn))(mean)
```

**Comment by CZ-Wu** [comment]
Here are my experiments with the correct jacobian (grey) and the wrong jacobian (orange).
![zipnerf](https://user-images.githubusercontent.com/27475044/236608302-dd3a8d7e-db2b-4a68-9f9b-f3598c29f245.png)

**Comment by SuLvXiangXin** [comment]
@CZ-Wu Thanks for  your attention, there are some mistakes in the first commit. But when I check it in the latest commit, and I find it allclose with the pytorch official `jacrev`. 
```
def track_linearize(fn, mean, std):
    """Apply function `fn` to a set of means and covariances, ala a Kalman filter.

  We can analytically transform a Gaussian parameterized by `mean` and `cov`
  with a function `fn` by linearizing `fn` around `mean`, and taking advantage
  of the fact that Covar[Ax + y] = A(Covar[x])A^T (see
  https://cs.nyu.edu/~roweis/notes/gaussid.pdf for details).

  Args:
    fn: the function applied to the Gaussians parameterized by (mean, cov).
    mean: a tensor of means, where the last axis is the dimension.
    std: a tensor of covariances, where the last two axes are the dimensions.

  Returns:
    fn_mean: the transformed means.
    fn_cov: the transformed covariances.
  """
    if fn == 'contract':
        fn = contract_mean_jacobi
    else:
        raise NotImplementedError

    pre_shape = mean.shape[:-1]
    mean = mean.reshape(-1, 3)
    std = std.reshape(-1)

    jvp_1, mean_1 = vmap(jacrev(contract_tuple, has_aux=True))(mean)
    std_1 = std * torch.linalg.det(jvp_1) ** (1 / mean.shape[-1])

    mean_2, jvp_2 = fn(mean)
    std_2 = std * torch.linalg.det(jvp_2) ** (1 / mean.shape[-1])

    mean_3, std_3 = contract_mean_std(mean, std)  # calculate det explicitly by using eigenvalues
    torch.allclose(std_1, std_3, atol=1e-7)  # True
    torch.allclose(mean_1, mean_3)  # True
    import ipdb; ipdb.set_trace()
    mean, std = contract_mean_std(mean, std)  # calculate det explicitly by using eigenvalues

    mean = mean.reshape(*pre_shape, 3)
    std = std.reshape(*pre_shape)
    return mean, std

```

Can you provide the script that how you test it? It'll be appreciated.

**Comment by CZ-Wu** [comment]
@SuLvXiangXin Thanks for your reply. The jacobian of your latest version is correct. I have figured out that your implementation uses the decorator of `@torch.no_grad()` which stops the gradients to my pose refine module, so the performance drops a little. Sorry for disturbing!

**Comment by Ilyabasharov** [comment]
@CZ-Wu how did you solve problem with pose refinement with `@torch.no_grad()` ? I want to apply pose optimisation too
